// @daostack/arc.js/dist/test/wrappers/testWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commonTypes_1 = require("../../commonTypes");
const contractWrapperBase_1 = require("../../contractWrapperBase");
const contractWrapperFactory_1 = require("../../contractWrapperFactory");
const web3EventService_1 = require("../../web3EventService");
class TestWrapperWrapper extends contractWrapperBase_1.ContractWrapperBase {
    constructor() {
        super(...arguments);
        this.name = "AbsoluteVote";
        this.friendlyName = "Test Wrapper";
        this.factory = exports.TestWrapperFactory;
    }
    foo() {
        return "bar";
    }
    aMethod() {
        return "abc";
    }
    setParameters(params) {
        params = Object.assign({}, {
            ownerVote: true,
            reputation: "0x1000000000000000000000000000000000000000",
            votePerc: 50,
        }, params);
        return super._setParameters("AbsoluteVote.setParameters", params.txEventContext, params.reputation, params.votePerc, params.ownerVote);
    }
    getDefaultPermissions() {
        return commonTypes_1.DefaultSchemePermissions.MinimumPermissions;
    }
}
exports.TestWrapperWrapper = TestWrapperWrapper;
exports.TestWrapperFactory = new contractWrapperFactory_1.ContractWrapperFactory("AbsoluteVote", TestWrapperWrapper, new web3EventService_1.Web3EventService());
